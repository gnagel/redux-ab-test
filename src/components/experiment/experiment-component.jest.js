import React from 'react'; // eslint-disable-line no-unused-vars
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { spy } from 'sinon';
import { Experiment } from './experiment';
import Variation from '../variation';
import reduxAbTest, { initialState } from '../../module';

const reduxAbTestState = initialState.merge({
  experiments: [
    {
      name:       'Test-experimentName',
      variations: [
        { name: 'Original', weight: 10000 },
        { name: 'Variation B', weight: 0 },
      ],
    },
  ],
  'availableExperiments': {
    'Test-experimentName': 'Test-experimentName',
  },
  'active': { 'Test-experimentName': 'Variation B' },
});

describe('(Component) src/components/experiment/experiment.js', () => {
  let component;
  let tree;
  let props;
  let dispatchActivate;
  let dispatchDeactivate;
  let dispatchPlay;
  let dispatchWin;

  beforeEach(() => {
    dispatchActivate = spy();
    dispatchDeactivate = spy();
    dispatchPlay = spy();
    dispatchWin = spy();

    props = {
      reduxAbTest: reduxAbTestState,
      name:     'Test-experimentName',
      children: [
        <Variation name="Original">Test Original</Variation>,
        <Variation name="Variation B">Test Variation B</Variation>,
      ],
      experiment: reduxAbTestState.getIn(['experiments', 0]),
      variation:  reduxAbTestState.getIn(['experiments', 0, 'variations', 1]),
      dispatchActivate,
      dispatchDeactivate,
      dispatchPlay,
      dispatchWin,
    };
    component = renderer.create(
      <Provider store={createStore(reduxAbTest, { reduxAbTest: reduxAbTestState })}>
        <Experiment {...props} />
      </Provider>
    );
    tree = component.toJSON();
  });

  it('exists', () => {
    expect(component).not.toBeUndefined;
  });

  it('has 1x rendered Experiment', () => {
    expect(tree).toMatchSnapshot();
    expect(tree.type).toEqual('span');
    expect(tree.children).toEqual(['Test Variation B']);
  });

  it('has 1x rendered variation', () => {
    expect(tree).toMatchSnapshot();
    expect(tree.type).toEqual('span');
    expect(tree.props['data-experiment-id']).toBeUndefined;
    expect(tree.props['data-experiment-name']).toEqual('Test-experimentName');
    expect(tree.props['data-variation-id']).toBeUndefined;
    expect(tree.props['data-variation-name']).toEqual('Variation B');
    expect(tree.children).toEqual(['Test Variation B']);
  });

  it('wraps the text children in a Variation', () => {
    props['children'] = 'Testing a single child';
    component = renderer.create(
      <Provider store={createStore(reduxAbTest, { reduxAbTest: reduxAbTestState })}>
        <Experiment {...props} />
      </Provider>
    );
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    expect(tree.type).toEqual('span');
    expect(tree.props['data-experiment-id']).toBeUndefined;
    expect(tree.props['data-experiment-name']).toEqual('Test-experimentName');
    expect(tree.props['data-variation-id']).toBeUndefined;
    expect(tree.props['data-variation-name']).toEqual('Variation B');
    expect(tree.children).toEqual(['Testing a single child']);
  });

  it('wraps the component children in a Variation', () => {
    props['children'] = <div>Testing a single child</div>;
    component = renderer.create(
      <Provider store={createStore(reduxAbTest, { reduxAbTest: reduxAbTestState })}>
        <Experiment {...props} />
      </Provider>
    );
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    expect(component).not.toBeUndefined;
    expect(tree.type).toEqual('div');
    expect(tree.props['data-experiment-id']).toBeUndefined;
    expect(tree.props['data-experiment-name']).toEqual('Test-experimentName');
    expect(tree.props['data-variation-id']).toBeUndefined;
    expect(tree.props['data-variation-name']).toEqual('Variation B');
    expect(tree.children).toEqual(['Testing a single child']);
  });

  it('renders null on blank children', () => {
    props['children'] = [];
    component = renderer.create(
      <Provider store={createStore(reduxAbTest, { reduxAbTest: reduxAbTestState })}>
        <Experiment {...props} />
      </Provider>
    );
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    expect(component).not.toBeUndefined;
    expect(tree).toBeNull;
  });

  it('creates Ad-hoc experiments');

  it('should update on componentWillReceiveProps');

  describe.skip('find by :id', () => {
    const reduxAbTestState = initialState.merge({
      experiments: [
        {
          id:         'Test-id',
          name:       'Test-experimentName',
          variations: [
            { name: 'Original', weight: 10000 },
            { name: 'Variation B', weight: 0 },
          ],
        },
      ],
      'availableExperiments': { 'Test-id': 'Test-id' },
      'active':               { 'Test-id': 'Variation B' },
    });

    beforeEach(() => {
      props = {
        reduxAbTest: reduxAbTestState,
        id:       'Test-id',
        children: [
          <Variation name="Original">Test Original</Variation>,
          <Variation name="Variation B">Test Variation B</Variation>,
        ],
        dispatchActivate,
        dispatchDeactivate,
        dispatchPlay,
        dispatchWin,
      };
      component = renderer.create(
        <Provider store={createStore(reduxAbTest, { reduxAbTest: reduxAbTestState })}>
          <Experiment {...props} />
        </Provider>
      );
      tree = component.toJSON();
    });

    it('exists', () => {
      expect(component).not.toBeUndefined;
      expect(component.html()).to.be.present;
    });

    it('has 1x rendered Experiment', () => {
      expect(component.find(Experiment)).to.have.length(1);
      expect(tree.type).toEqual('span');
      expect(tree.children).toEqual('Test Variation B');
    });
  });

  describe.skip('find by :selector', () => {
    const reduxAbTestState = initialState.merge({
      experiments: [
        {
          id:        'Test-id',
          name:      'Test-experimentName',
          component: {
            key: 'Example component key selector',
          },
          variations: [
            { name: 'Original', weight: 10000 },
            { name: 'Variation B', weight: 0 },
          ],
        },
      ],
      availableExperiments: {
        'Example component key selector': 'Test-id',
      },
      active:   { 'Test-id': 'Variation B' },
      key_path: ['component', 'key'],
    });

    beforeEach(() => {
      props = {
        reduxAbTest: reduxAbTestState,
        selector: 'Example component key selector',
        children: [
          <Variation name="Original">Test Original</Variation>,
          <Variation name="Variation B">Test Variation B</Variation>,
        ],
        dispatchActivate,
        dispatchDeactivate,
        dispatchPlay,
        dispatchWin,
      };
      component = renderer.create(
        <Provider store={createStore(reduxAbTest, { reduxAbTest: reduxAbTestState })}>
          <Experiment {...props} />
        </Provider>
      );
      tree = component.toJSON();
    });

    it('exists', () => {
      expect(component).not.toBeUndefined;
      expect(component.html()).to.be.present;
    });

    it('has 1x rendered Experiment', () => {
      expect(component.find(Experiment)).to.have.length(1);
      expect(tree.type).toEqual('span');
      expect(tree.children).toEqual('Test Variation B');
    });

  });
});
