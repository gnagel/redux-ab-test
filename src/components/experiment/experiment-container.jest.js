import React from 'react'; // eslint-disable-line no-unused-vars
import Immutable from 'immutable';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { spy } from 'sinon';
import Experiment, { mapStateToProps, mapDispatchToProps } from './experiment';
import Variation from '../variation';
import reduxAbTest, { initialState } from '../../module';


describe('(Container) Experiment', () => {
  describe.skip('mapStateToProps', () => {
    it('exists', () => {
      expect(mapStateToProps).not.toBeUndefined;
    });

    it('has correct keys', () => {
      expect(Object.keys(mapStateToProps({reduxAbTest: initialState}, {}))).toEqual(['reduxAbTest', 'experiment', 'variation']);
    });

    it('has correct state', () => {
      const props = {
        name: 'Test-experimentName',
      };
      const store = {
        reduxAbTest: initialState.merge({
          experiments: [
            {
              name:       'Test-experimentName',
              variations: [
                { name: 'Original', weight: 10000 },
                { name: 'Variation B', weight: 0 },
              ],
            },
          ],
          availableExperiments: {
            'Test-experimentName': 'Test-experimentName',
          },
          active: { 'Test-experimentName': 'Variation B' },
        }),
      };
      const output = mapStateToProps(store, props);
      expect(Object.keys(output)).toEqual(['reduxAbTest', 'experiment', 'variation']);
      const experiment = output.experiment;
      const variation = output.variation;
      expect(experiment).not.toBeNull;
      expect(variation).not.toBeNull;
      expect(experiment.toJS()).toEqual(store.reduxAbTest.getIn(['experiments', 0]).toJS());
      expect(variation.toJS()).toEqual(store.reduxAbTest.getIn(['experiments', 0, 'variations', 1]).toJS());
    });


    it('has correct experiment', () => {
      expect(Object.keys(mapStateToProps({reduxAbTest: initialState}, {}))).toEqual(['reduxAbTest', 'experiment', 'variation']);
    });
  });

  describe.skip('mapDispatchToProps', () => {
    it('exists', () => {
      expect(mapDispatchToProps).not.toBeUndefined;
    });
    it('has correct keys', () => {
      const dispatch = spy();
      expect(Object.keys(mapDispatchToProps(dispatch))).toEqual(['dispatchActivate', 'dispatchDeactivate', 'dispatchPlay', 'dispatchWin']);
    });
  });

  it.skip('renders something', () => {
    const experiment = Immutable.fromJS({
      name:       'Test-experimentName',
      id:         'Test-id',
      variations: [
        { name: 'Original', weight: 10000 },
        { name: 'Variation B', weight: 0 },
      ],
    });
    const state = {
      reduxAbTest: initialState.merge({
        experiments: [ experiment ],
        availableExperiments: {
          'Test-experimentName': 'Test-id',
        },
      }),
    };
    const component = renderer.create(
      <Provider store={createStore(reduxAbTest, state)}>
        <Experiment
          selector='Test-experimentName'
          experiment={experiment}
          variation={experiment.getIn(['variations', 1])}
          dispatchActivate={() => { console.log('dispatchActivate!'); }}
          dispatchDeactivate={() => { console.log('dispatchDeactivate!'); }}
          dispatchPlay={() => { console.log('dispatchPlay!'); }}
          dispatchWin={() => { console.log('dispatchWin!'); }}
          >
          <Variation
            name='Original'
            >
            Original Version
          </Variation>
          <Variation
            name='Variation B'
            >
            B Version
          </Variation>
        </Experiment>
      </Provider>
    );
    expect(component).not.toBeUndefined;
  });

  describe.skip('component', () => {
    let component;
    let tree;

    beforeEach(() => {
      const state = {
        reduxAbTest: initialState.merge({
          experiments: [
            {
              name:       'Test-experimentName',
              variations: [
                { name: 'Original', weight: 10000 },
                { name: 'Variation B', weight: 0 },
              ],
            },
          ],
          availableExperiments: {
            'Test-experimentName': 'Test-experimentName',
          },
        }),
      };
      const props = {
        reduxAbTest: state.reduxAbTest,
        name:     'Test-experimentName',
        children: [
          <Variation name="Original">Test Original</Variation>,
          <Variation name="Variation B">Test Variation B</Variation>,
        ],
      };
      component = renderer.create(
        <Provider store={createStore(reduxAbTest, state)}>
          <Experiment {...props} />
        </Provider>
      );
      tree = component.toJSON();
    });

    it.skip('exists', () => {
      expect(component).not.toBeUndefined;
      expect(tree).toMatchSnapshot();
    });

    it.skip('has 1x Variation', () => {
      expect(tree).toMatchSnapshot();
      expect(tree.type).toEqual('span');
      expect(tree.props['data-experiment-id']).toBeUndefined;
      expect(tree.props['data-experiment-name']).toEqual('Test-experimentName');
      expect(tree.props['data-variation-id']).toBeUndefined;
      expect(tree.props['data-variation-name']).toEqual('Variation B');
      expect(tree.children).toEqual(['Test Variation B']);
    });

    describe.skip('find by :id', () => {
      let props;
      let dispatchActivate;
      let dispatchDeactivate;
      let dispatchPlay;
      let dispatchWin;

      const reduxAbTestState = initialState.merge({
        experiments: [
          {
            id:         'Test-id',
            name:       'Test-experimentName',
            variations: [
              { name: 'Original', weight: 10000 },
              { name: 'Variation B', weight: 0 },
            ],
          },
        ],
        'availableExperiments': { 'Test-id': 'Test-id' },
        'active':               { 'Test-id': 'Variation B' },
      });

      beforeEach(() => {
        dispatchActivate = spy();
        dispatchDeactivate = spy();
        dispatchPlay = spy();
        dispatchWin = spy();

        props = {
          reduxAbTest: reduxAbTestState,
          id:       'Test-id',
          children: [
            <Variation name="Original">Test Original</Variation>,
            <Variation name="Variation B">Test Variation B</Variation>,
          ],
          dispatchActivate,
          dispatchDeactivate,
          dispatchPlay,
          dispatchWin,
        };
        component = renderer.create(
          <Provider store={createStore(reduxAbTest, { reduxAbTest: reduxAbTestState })}>
            <Experiment {...props} />
          </Provider>
        );
        tree = component.toJSON();
      });

      it('exists', () => {
        expect(component).not.toBeUndefined;
        expect(component.html()).to.be.present;
      });

      it('has 1x rendered Experiment', () => {
        expect(component.find(Experiment)).to.have.length(1);
        expect(tree.type).toEqual('span');
        expect(tree.children).toEqual('Test Variation B');
      });
    });

    describe.skip('find by :selector', () => {
      let props;
      let dispatchActivate;
      let dispatchDeactivate;
      let dispatchPlay;
      let dispatchWin;

      const reduxAbTestState = initialState.merge({
        experiments: [
          {
            id:        'Test-id',
            name:      'Test-experimentName',
            component: {
              key: 'Example component key selector',
            },
            variations: [
              { name: 'Original', weight: 10000 },
              { name: 'Variation B', weight: 0 },
            ],
          },
        ],
        availableExperiments: {
          'Example component key selector': 'Test-id',
        },
        active:   { 'Test-id': 'Variation B' },
        key_path: ['component', 'key'],
      });

      beforeEach(() => {
        dispatchActivate = spy();
        dispatchDeactivate = spy();
        dispatchPlay = spy();
        dispatchWin = spy();

        props = {
          reduxAbTest: reduxAbTestState,
          selector: 'Example component key selector',
          children: [
            <Variation name="Original">Test Original</Variation>,
            <Variation name="Variation B">Test Variation B</Variation>,
          ],
          dispatchActivate,
          dispatchDeactivate,
          dispatchPlay,
          dispatchWin,
        };
        component = renderer.create(
          <Provider store={createStore(reduxAbTest, { reduxAbTest: reduxAbTestState })}>
            <Experiment {...props} />
          </Provider>
        );
        tree = component.toJSON();
      });

      it('exists', () => {
        expect(component).not.toBeUndefined;
        expect(component.html()).to.be.present;
      });

      it('has 1x rendered Experiment', () => {
        expect(component.find(Experiment)).to.have.length(1);
        expect(tree.type).toEqual('span');
        expect(tree.children).toEqual('Test Variation B');
      });

    });
  });
});
