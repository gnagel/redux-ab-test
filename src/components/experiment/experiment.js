/** @flow */
import React from 'react';
import Immutable from 'immutable';
import { isVariationComponent } from '../../components/variation';
import Variation, { isVariationContainer } from '../../containers/variation';
import selectVariation from '../../utils/select-variation';
import { cacheStore } from '../../utils/create-cache-store';
import getKey from '../../utils/get-key';

type Props = {
  /**
   * Redux State:
   * ```js
   *   reduxAbTest = {
   *     experiments: [],
   *     active: {
   *       "experiment.name" => "variation.name",
   *       ...
   *     }
   *     ...
   *   }
   * ```
   */
  reduxAbTest: Immutable.Map,
  /**
   * Selector key for choosing an experiment from the redux store.
   * This can be used as such:
   *  <Experiment selector=`Homepage Header Experiment`> ... </Experiment>
   */
  selector: ?string,
  /**
   * ID of the experiment, this is used to pick a specific experient from the redux store.
   * We use getKey to generate the `key` for this experiment.
   */
  id: ?string,
  /**
   * Name of the experiment, this is used to pick a specific experient from the redux store.
   * This should only be used as a fallback if `id` is unavailable
   */
  name: string,
  /**
   * Name of the default variation.
   * >  When defined, this value is used to choose a variation if a stored value is not present.
   * >  This property may be useful for server side rendering but is otherwise not recommended.
   */
  defaultVariationName: ?string,
  /**
   * Action Creator callback: Function({experiment:ExperimentType})
   */
  dispatchActivate: Function,
  /**
   * Action Creator callback: Function({experiment:ExperimentType})
   */
  dispatchDeactivate: Function,
  /**
   * Action Creator callback: Function({experiment:ExperimentType, variation:VariationType})
   */
  dispatchPlay: Function,
  /**
   * Action Creator callback: Function({experiment:ExperimentType, variation:VariationType})
   */
  dispatchWin: Function,
  /**
   * Variation Children to render
   */
  children: any,
};


type State = {
  /**
   * The currenly active experiment
   */
  experiment: ?ExperimentType,
  /**
   * The currenly active variation
   */
  variation: ?VariationType,
  /**
   * The experient was played
   */
  played: ?bool,
};

const isValid = (child) => (isVariationContainer(child) || isVariationComponent(child));


export default class Experiment extends React.Component {
  props: Props;
  state: State;
  static defaultProps = {
    reduxAbTest:          Immutable.Map({}),
    id:                   null,
    name:                 null,
    selector:             null,
    defaultVariationName: null,
    dispatchActivate:     () => {},
    dispatchDeactivate:   () => {},
    dispatchPlay:         () => {},
    dispatchWin:          () => {},
  };
  state = {
    variation:  null,
    experiment: null,
    played:     false,
  };


  /**
   * Activate the variation
   */
  componentWillMount() {
    const { id, name, selector, defaultVariationName, reduxAbTest, dispatchActivate, dispatchPlay } = this.props;
    const experiment = this.getExperiment(reduxAbTest, selector, id, name);

    // If the experiment is unavailable, then record it wasn't played and move on
    if (!experiment) {
      this.setState({ experiment: null, variation: null, played: false });
      return;
    }

    const variation = selectVariation({
      active: reduxAbTest.get('active'),
      experiment,
      defaultVariationName,
      cacheStore
    });

    // These will trigger `componentWillReceiveProps`
    dispatchActivate({experiment});
    dispatchPlay({experiment, variation});
    this.setState({ experiment, variation, played: true });
  }


  /**
   * Update the component's state with the new properties
   */
  componentWillReceiveProps(nextProps:Props) {
    const { selector, id, name, defaultVariationName, reduxAbTest, dispatchActivate, dispatchPlay } = nextProps;
    const experiment = this.getExperiment(reduxAbTest, selector, id, name);
    if (!experiment) {
      // If we no-longer have an experiment anymore, then update the internal state
      if (this.state.experiment) {
        this.setState({ experiment: null, variation: null, played: false });
      }
      return;
    }

    const variation = selectVariation({
      active: reduxAbTest.get('active'),
      experiment,
      defaultVariationName,
      cacheStore
    });

    if (!experiment.equals(this.state.experiment) || !variation.equals(this.state.variation)) {
      // These will trigger `componentWillReceiveProps`
      if (!this.state.played) {
        dispatchActivate({experiment});
        dispatchPlay({experiment, variation});
      }
      this.setState({ experiment, variation, played: true });
    }
  }

  getExperiment(reduxAbTest, selector, id, name) {
    // Find the key of the currently available experiment
    const key = reduxAbTest.getIn(['availableExperiments', selector || id || name], null);
    if (!key) {
      return null;
    }
    // Select the experiment from the redux store
    const experiment = reduxAbTest.get('experiments').find(experiment => (getKey(experiment) === key), null);
    // Return the resulting experiment
    return experiment;
  }

  /**
   * Deactivate the variation from the state
   */
  componentWillUnmount() {
    const { dispatchDeactivate } = this.props;
    const { experiment } = this.state;
    // Dispatch the deactivation event
    if (experiment) {
      dispatchDeactivate({experiment});
    }
  }


  /**
   * Render one of the variations or `null`
   */
  render() {
    const { children, defaultVariationName } = this.props;
    const { experiment } = this.state;
    const variation = this.state.variation || Immutable.Map({ name: defaultVariationName, id: null });

    const childrenArray = React.Children.toArray(children);

    // If there are no children, render nothing
    if (childrenArray.length === 0) {
      return null;
    }

    // If the first child is text or an unknown component, simply wrap it in a Variation
    if (childrenArray.length === 1 && !isValid(childrenArray[0])) {
      return (
        <Variation id={variation.get('id')} name={variation.get('name')} experiment={experiment} variation={variation}>
          {children}
        </Variation>
      );
    }

    const selectedChild = [
      childrenArray.find( child => (variation.get('id')   && child.props.id   === variation.get('id'))   ),
      childrenArray.find( child => (variation.get('name') && child.props.name === variation.get('name')) ),
    ].filter( value => value ).find( value => value );
    if (!selectedChild) {
      throw new Error(`Expected to find a Variation child matching id=${variation.get('id')} or name=${variation.get('name')}`);
    }

    // Inject the helper `handleWin` into the child element
    return React.cloneElement(selectedChild, {
      experiment,
      variation,
      id:   variation.get('id'),
      name: variation.get('name'),
    });
  }
}
