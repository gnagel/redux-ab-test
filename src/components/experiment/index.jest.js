import Experiment from './index';

describe('(Component) src/components/experiment/index.js', () => {
  it('Experiment exists', () => {
    expect(Experiment).not.toBeUndefined;
  });
});
