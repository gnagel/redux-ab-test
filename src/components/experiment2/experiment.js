import React                  from 'react';
import Immutable              from 'immutable';
import { connect }            from 'react-redux';
import { bindActionCreators } from 'redux';
import {
  play,
  activate,
  deactivate,
}                             from '../../module';
import Variation              from './variation';
import { logger }             from './logger';


class Experiment extends React.Component {
  render() {
    const { name, children } = this.props;
    logger(`Rendering Experiment name='${name}'`);

    // Group the children by name
    const childrenByName = this.groupChildrenByName();
    logger(`Experiment name='${name}' has children with names='${Object.keys(childrenByName)}'`);

    // Lookup the experiment and variation for this experiment
    const exeriment = this._experiment || this.getExperiment();
    const variation = this._variation || this.getVariation(experiment);

    // Find the attached exeriment & variation for this component
    if (!exeriment) {
      logger(`Experiment name='${name}' didn't have a valid exeriment`);
      return null;
    }
    if (!variation) {
      logger(`Experiment name='${name}' didn't have a valid variation`);
      return null;
    }
    // Find the variation child with the matching name
    const variationName = variation.get('name');
    const child = childrenByName[variationName] || null;
    if (!child) {
      logger(`Experiment name='${name}' found no variation matching variation.name='${variationName}'`);
      return null;
    }
    // Return the rendered child
    logger(`Rendered Experiment name='${name}', variation.name='${variationName}'`);
    return child;
  }

  //
  // Helpers:
  //

  groupChildrenByName = () => {
    const childrenByName = {};
    if (React.Children.count(children) === 0) {
      childrenByName[ children.props.name ] = children;
    } else {
      React.Children.map(child => {
        child[child.props.name] = child;
      });
    }
    return childrenByName;
  };

  groupExperimentsByName = () => {
    const { reduxAbTest } = this.props;
    const experiments = reduxAbTest.getIn(['experiments'], Immutable.List());
    const experimentsByName = {};
    experiments.forEach(e => {
      experimentsByName[e.get('name', '').toLowerCase()] = e;
    });
    return experimentsByName;
  };

  getExperiment = () => {
    const { name } = this.props;
    return this.groupExperimentsByName()[name];
  };

  getVariation = (experiment) => {
    if (!experiment) {
      return null;
    }
    const variations = {};
    experiment.get('variations', Immutable.List()).forEach(variation => {
      variations[variation.get('name', '')] = variation;
    });
    return variations;
  };

  //
  // Component Lifecycle
  //

  // Activate the variation
  componentWillMount() {
    const { play } = this.props;
    const experiment = this.experiment();
    const variation = this.variation();
    // Record the experiment was shown
    if (play && experiment && variation) {
      play({ experiment, variation });
    }
    this._experiment = experiment;
    this._variation = variation;
  }

  // Activate the variation
  componentDidMount() {
    const { activate } = this.props;
    const experiment = this.experiment();
    if (activate && experiment) {
      activate({ experiment });
    }
    this._experiment = null;
    this._variation = null;
  }

  // Dispatch the deactivation event
  componentWillUnmount() {
    const { deactivate } = this.props;
    const experiment = this.experiment();
    if (deactivate && experiment) {
      deactivate({ experiment });
    }
    this._experiment = null;
    this._variation = null;
  }
}


export const requireChildType = (componentName, index, child) => {
  if (child.type === Variation) {
    return undefined;
  }
  return new Error(`'${componentName}' should have a children of the type: 'Variation': child.index='${index}', child.type='${child.type}'`);
};


export const validateChildren = (props, propName, componentName) => {
  const children = props[propName];
  // Only accept children of the appropriate type
  const childrenCount = React.Children.count(children);
  switch(childrenCount) {
  case 0:
    return new Error(`'${componentName}' should have at least one child of the type: 'Variation': children.count='${childrenCount}'`);
  case 1:
    return requireChildType(componentName, 0, children);
  default: {
    const errors = React.Children.map( (index, child) => validateChildren(componentName, index, child) ).filter( err => err );
    if (errors.length !== 0) {
      return errors[0];
    }
    return undefined;
  }
  }
};


Experiment.propTypes = {
  // Name of the experiment
  name:       React.PropTypes.string.isRequired,
  // Selector for the experiment
  isEnabled:  React.PropTypes.func.isRequired,
  // React element to render:
  children:   validateChildren,
  // Bound Action Creators
  play:       React.PropTypes.func.isRequired,
  win:        React.PropTypes.func.isRequired,
  activate:   React.PropTypes.func.isRequired,
  deactivate: React.PropTypes.func.isRequired,
};


export const mapStateToProps = ({ reduxAbTest }) => ({ reduxAbTest });
export const mapDispatchToProps = dispatch => bindActionCreators({ play, activate, deactivate }, dispatch);
export default connect(mapStateToProps, mapDispatchToProps)((props) => {
  // If not enabled, abort now
  const enabled = props.isEnabled();
  if (!enabled) {
    logger(`Experiment name='${props.name}' is not enabled`);
    return null;
  } else {
    logger(`Experiment name='${props.name}' is enabled`);
    return <Experiment {...props} />;
  }
});
