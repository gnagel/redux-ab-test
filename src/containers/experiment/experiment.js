/** @flow */
import React from "react"; // eslint-disable-line no-unused-vars
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { Experiment } from '../../components';
import { registerAdhoc, activate, deactivate, play, win } from '../../module';


/**
 * Map the Redux Store to the Experiment's props
 */
export const mapStateToProps = (state) => {
  const { reduxAbTest } = state;
  return { reduxAbTest };
};


/**
 * Map the action creators to the the Experiment's props.
 */
export const mapDispatchToProps = (dispatch) => {
  return {
    dispatchRegisterAdhoc: bindActionCreators(registerAdhoc, dispatch),
    dispatchActivate:      bindActionCreators(activate, dispatch),
    dispatchDeactivate:    bindActionCreators(deactivate, dispatch),
    dispatchPlay:          bindActionCreators(play, dispatch),
    dispatchWin:           bindActionCreators(win, dispatch)
  };
};


/**
 * Export the new React Container.
 */
export default connect(mapStateToProps, mapDispatchToProps)(Experiment);
