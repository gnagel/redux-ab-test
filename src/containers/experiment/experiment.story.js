import React from 'react'; //eslint-disable-line
import { storiesOf } from '@kadira/storybook';
import { decorateStore } from '../../../.storybook/store';
import Experiment from './experiment'; //eslint-disable-line
import Variation from '../variation'; //eslint-disable-line
import Debugger from '../debugger'; //eslint-disable-line

const __state__ = {
  experiments: [
    {
      id:         'Test-experimentId',
      name:       'Test-experimentName',
      variations: [
        { name: 'Original', weight: 10000 },
        { name: 'Variation B', weight: 0 }
      ],
    },
  ],
};


storiesOf('Experiment', module)
  .addDecorator(decorateStore(__state__))

  .add('Original variation in an experiment', () => (
    <div>
      <Experiment
        name='Test-experimentName'
        experimentName='Test-experimentName'
        >
        <Variation name='Original'>Original content</Variation>
        <Variation name='Variation B'>Variation B's content</Variation>
      </Experiment>
      <Debugger />
    </div>
  ))
  ;
