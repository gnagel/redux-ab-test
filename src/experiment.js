/** @flow */
import React from "react";
import Immutable from 'immutable';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux'

import Variation from "./variation";
import { ExperimentType, VariationType, actions, selectors } from './module';


type Props = {
  /**
   * Name of the experiment
   */
  name: string,
  /**
   * Name of the default variation.
   * >  When defined, this value is used to choose a variation if a stored value is not present.
   * >  This property may be useful for server side rendering but is otherwise not recommended.
   */
  defaultVariationName: ?string,
  /**
   * Distinct user identifier.
   * >  When defined, this value is hashed to choose a variation
   * >  if `defaultVariationName` or a stored value is not present.
   * >  Useful for server side rendering.
   */
  userIdentifier: ?string,
  /**
   * Redux State:
   * ```js
   *   reduxAbTest = {
   *     experiments: [],
   *     active: {
   *       "experiment.name" => "variation.name",
   *       ...
   *     }
   *     ...
   *   }
   * ```
   */
  reduxAbTest: Immutable.Map,
  /**
   * Bound Action Creator: actions.activate(experimentName:string)
   */
  dispatchActivate: Function,
  /**
   * Bound Action Creator: actions.deactivate(experimentName:string)
   */
  dispatchDeactivate: Function,
  /**
  * Bound Action Creator: actions.play(experimentName:string, variationName:string)
   */
  dispatchPlay: Function,
  /**
   * Bound Action Creator: actions.win(experimentName:string, variationName:string)
   */
  dispatchWin: Function,
};


type State = {
  /**
   * Hash of "name" => Variation element
   */
  variationElements: Object,
  /**
   * The currenly active experiment
   */
  experiment: ?ExperimentType,
  /**
   * The currenly active variation
   */
  variation: ?VariationType,
};


export default class Experiment extends React.Component {
  props: Props;
  state: State;

  constructor(props, context) {
    super(props, context);
    this.state = {
      variationElements: {},
      variation: null,
      experiment: null
    };
  }

  /**
   * Activate the variation
   */
  componentWillMount() {
    const { name, defaultVariationName, userIdentifier, reduxAbTest, dispatchActivate, dispatchPlay, children } = this.props;
    const experiment = selectors.findExperiment({reduxAbTest, experimentName: name});
    const variationElements = selectors.mapChildrenToVariationElements(children);
    const variation = selectors.selectVariation({
      experiment,
      defaultVariationName,
      userIdentifier,
      reduxAbTest
    });

    // These will trigger `componentWillReceiveProps`
    dispatchActivate(name, variation);
    dispatchPlay(name, variation);
    // Update the state
    this.setState({ variationElements, experiment, variation });
  }

  /**
   * Update the component's state with the new properties
   */
  componentWillReceiveProps(nextProps) {
    const { name, defaultVariationName, userIdentifier, reduxAbTest, children } = nextProps;
    const experiment = selectors.findExperiment({reduxAbTest, experimentName: name});
    const variationElements = selectors.mapChildrenToVariationElements(children);
    const variation = selectors.selectVariation({
      experiment,
      defaultVariationName,
      userIdentifier,
      reduxAbTest
    });
    this.setState({ variationElements, experiment, variation });
  }

  /**
   * Deactivate the variation from the state
   */
  componentWillUnmount() {
    const { dispatchDeactivate } = this.props;
    const { experiment } = this.state;
    dispatchDeactivate(experiment);
  }

  /**
   * Render one of the variations or `null`
   */
  render() {
    const { experiment, variation, variationElements } = this.state;
    return variationElements[variation.name] || null;
  }
}


export const mapStateToProps = (state) => {
  const { reduxAbTest } = state;
  return { reduxAbTest };
}

export const mapDispatchToProps = (dispatch) => {
  return {
    dispatchActivate: bindActionCreators(actions.activate, dispatch),
    dispatchDeactivate: bindActionCreators(actions.deactivate, dispatch),
    dispatchPlay: bindActionCreators(actions.play, dispatch),
    dispatchWin: bindActionCreators(actions.win, dispatch),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Experiment);
