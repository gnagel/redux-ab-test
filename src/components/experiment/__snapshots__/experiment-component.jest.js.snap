// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) src/components/experiment/experiment.js has 1x rendered Experiment 1`] = `
<span
  data-experiment-id={undefined}
  data-experiment-name="Test-experimentName"
  data-variation-id={undefined}
  data-variation-name="Variation B"
>
  Test Variation B
</span>
`;

exports[`(Component) src/components/experiment/experiment.js has 1x rendered variation 1`] = `
<span
  data-experiment-id={undefined}
  data-experiment-name="Test-experimentName"
  data-variation-id={undefined}
  data-variation-name="Variation B"
>
  Test Variation B
</span>
`;

exports[`(Component) src/components/experiment/experiment.js renders null on blank children 1`] = `null`;

exports[`(Component) src/components/experiment/experiment.js wraps the component children in a Variation 1`] = `
<div
  data-experiment-id={undefined}
  data-experiment-name="Test-experimentName"
  data-variation-id={undefined}
  data-variation-name="Variation B"
>
  Testing a single child
</div>
`;

exports[`(Component) src/components/experiment/experiment.js wraps the text children in a Variation 1`] = `
<span
  data-experiment-id={undefined}
  data-experiment-name="Test-experimentName"
  data-variation-id={undefined}
  data-variation-name="Variation B"
>
  Testing a single child
</span>
`;
